src/matrixrtc/BasicEncryptionManager.ts

```diff
@@ -0,0 +1,237 @@
+     */
+    private ensureMediaKey(): void {
+        if (this.currentKeyDistributionPromise == null) {
+            this.logger.debug(`No active rollout, start a new one`);
+            // start a rollout
+            this.currentKeyDistributionPromise = this.rolloutOutboundKey().then(() => {
+                this.logger.debug(`Rollout completed`);
+                this.currentKeyDistributionPromise = null;
+                if (this.needToRotateAgain) {
+                    this.logger.debug(`New Rollout needed`);
+                    // rollout a new one
+                    this.ensureMediaKey();
+                }
+            });
+        } else {
+            // There is a rollout in progress, but membership has changed and a new rollout is needed.
+            // Remember that a new rotation in needed after the current one.
+            this.logger.debug(`Rollout in progress, a new rollout will be started after the current one`);
+            this.needToRotateAgain = true;
+        }
+    }
+
+    public leave(): void {
+        this.keyBuffer.clear();
+        this.transport.off(KeyTransportEvents.ReceivedKeys, this.onNewKeyReceived);
+        this.transport.stop();
+    }
+
+    public onNewKeyReceived: KeyTransportEventListener = (userId, deviceId, keyBase64Encoded, index, timestamp) => {
+        this.logger.debug(`Received key over transport ${userId}:${deviceId} at index ${index}`);
+
+        // We have a new key notify the video layer of this new key so that it can decrypt the frames properly.
+        // We also store a copy of the key in the key store as we might need to re-emit them to the decoding layer.
+        const participantId = getParticipantId(userId, deviceId);
+        const keyBin = decodeBase64(keyBase64Encoded);
+        const newKey: InboundEncryptionSession = {
+            key: keyBin,
+            participantId,
+            keyId: index,
+            creationTS: timestamp,
+        };
+
+        const validKey = this.keyBuffer.disambiguate(participantId, newKey);
+        if (validKey) {
+            this.onEncryptionKeysChanged(validKey.key, index, validKey.participantId);
+            this.statistics.counters.roomEventEncryptionKeysReceived += 1;
+        } else {
+            this.logger.info(`Received an out of order key for ${userId}:${deviceId}, dropping it`);
+        }
+    };
+
+    /**
+     * Called when the membership of the call changes.
+     * This encryption manager is very basic, it will rotate the key everytime this is called.
+     * @param oldMemberships
+     */
+    public onMembershipsUpdate(oldMemberships: CallMembership[]): void {
+        this.logger.trace(`onMembershipsUpdate`);
+
+        // This encryption manager is very basic, it will rotate the key for any membership change
+        // Request rotation of the key
+        this.ensureMediaKey();
+    }
+
+    private async rolloutOutboundKey(): Promise<void> {
+        const hasExistingKey = this.outboundSession != null;
```

Comment:
```suggestion
        const isFirstKey = this.outboundSession === null;
```
I think this makes the logic where its used easier to read